README.md

```diff
@@ -16,6 +16,8 @@ give you write access.
 
 #### Alternative/Derived implementations:
 
+- https://github.com/hedronvision/bazel-compile-commands-extractor
+  - Uses an alternate approach based on aquery. This directly asks Bazel for its exact build commands, while still being fast and not requiring a full build. Avoids whole categories of issues.
```

Comment:
@cpsauer Can you elaborate on "whole categories of issues" here? (we're trying to evaluate different solutions and would like to understand this)

Reply:
Hey, @thii! Good to hear from you again--it's been a bit since we interacted on the interwebs, but I still really appreciate your help with, I think, tulsi stuff, back when I was fixing some of the external workspace things there.

Sure; happy to. But it's been a little bit since I first wrote that, so I'm dredging up from memory. I do remember we prototyped implementing it a few different ways and found aquery to be the best by far. IIRC, the reasoning for aquery vs aspects/query is that, fundamentally the database is a listing of command line actions, which is direct mapping with aquery, rather than crawling the graph of targets you'd get with aspects/query. With aspects/query, there were lots of problems and additional lift required to convert things into commands correctly, whereas you get the exact command for free with aquery. In particular, you'd otherwise have to duplicate the attr->command line logic from bazel & the toolchain, and IIRC that gets pretty brittle with custom flags and toolchains and on other platforms, especially with in-graph configuration transitions for cross-compiling. Aquery gets all that basically for free by looking directly at the compile command, though you do need to do some light de-bazeling when Bazel use wrapper scripts around the compiler drivers, like on Apple platforms. Aquery can also easily filter to Cpp/Objc compile actions, even for actions that are for code-gen tools being built on the host, and I think that's not so easy with the other techniques. [We also evaluated--and built--a working version based on action_listeners, which are now deprecated by bazel, but that too was worse. Happy to explain more if you're curious, but I think you're probably evaluating vs the aspect/query approach done here.]

I'm curious what you're evaluating/building!

Reply:
Thanks for the context. Yes, we are evaluating aspects vs aquery. We haven't had the above issues with aspects, but definitely something to keep in mind. However, we found issues with with the aquery approach so far:
* Since aquery only does an analysis, you can't get correct compile commands for undetermined sources, e.g. when you have an input that is a tree artifact.
* The generated compilation database can reference to not-yet-existent source files (you need to do a full build to get those files). With aspects, you can request generated source files while generating the compilation database, without doing a full build.

Reply:
👍🏻 For this or for some other use case?
The latter might be solvable with a complementary aspect, at least. Any chance you have an example of the first for me to think about?

Reply:
```starlark
# rule.bzl
def _impl(ctx):
    dir_name = ctx.attr.name + ".c"
    out = ctx.actions.declare_directory(dir_name)
    ctx.actions.run(
        outputs = [out],
        executable = ctx.executable._tool,
        arguments = [
            ctx.bin_dir.path,
            dir_name,
        ],
    )
    return [DefaultInfo(files = depset([out]))]

codegen = rule(
    _impl,
    attrs = {
        "_tool": attr.label(
            executable = True,
            cfg = "exec",
            default = "//:tool.sh",
            allow_single_file = True,
        ),
    },
)
```
```starlark
# BUILD
load(":rule.bzl", "codegen")

codegen(name = "srcs")

cc_library(
    name = "a",
    srcs = [":srcs"],
)
```

```sh
# tool.sh
cd $1
mkdir -p $2
cd $2
touch inner-1.c
touch inner-2.c
```

Reply:
> For this or for some other use case?

Not sure what this question is about.

Reply:
^ I'm just curious what you're working on that's motivating the questions. Are you thinking about making a change to this repo--or building a different tool entirely, maybe for something else?

Reply:
I'll DM you on Slack.

Reply:
👍🏻 but could I get you to email me instead to dm? (Migrating computers--slack currently borked) christophersauer -at- pacbell.net

Reply:
@thii, bumping the above just bc I think there's a good chance I lost your message here